onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/Shaders/bluestein_chirp.hlsl

```diff
@@ -0,0 +1,77 @@
+// TBUFFER is the data type to read from src and write to dst.
+// Arithmetic is always done in FP32.
+#if !defined(TBUFFER)
+#define TBUFFER float
+#endif
+
+RWStructuredBuffer<TBUFFER> chirp : register(u0);
+RWStructuredBuffer<TBUFFER> b : register(u1);
+
+cbuffer Constants
+{
+    uint StartIndex;
+    uint ElementCount;
+    uint DFTLength;
+    uint IsInverse;
+};
+
+uint NextPowerOf2(uint x) {
+  x--;
+  uint y = 1;
+  while (y <= x) {
+    y <<= 1;
+  }
+  return y;
+}
+
+
+// Returns the indices for the real and complex output uav
+uint2 ComputeDestIndex(uint n)
+{
+    return uint2(n*2, n*2+1);
+}
+
+float2 CalculateChirp(uint n, uint N, bool isInverse)
+{
+    static const float PI = 3.14159265f;
+    float direction = isInverse ? 1 : -1;
+    // chirp[n] = e^(i * direction * pi * n * n / N)
+    float theta = direction * PI * n * n / N;
+    return float2(cos(theta), sin(theta));
+}
+
+[numthreads(64, 1, 1)]
+void DFT(uint3 dtid : SV_DispatchThreadId)
+{
+    uint n = StartIndex + dtid.x;
+    if (n < ElementCount)
+    {
+        uint N = DFTLength;
```

Comment:
Any reason we want to do this here? `DFTLength` is more self explanatory.

Reply:
Actually there are 2 DFT Lengths, one is the true DFT we want to calculate N, and the other is the synthetically enlarged DFT M which is a power of 2.